<a href="https://colab.research.google.com/github/vien-nguyenthimy/interrnship-ml-flyrank/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/vien-nguyenthimy/interrnship-ml-flyrank/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

I chose the Refresh / Content Opportunity Scoring lane

The goal is to help content or SEO teams identify which pages require review, particularly when time is limited. Instead of manually auditing pages one by one, I aim to leverage observable signals—such as search results, content characteristics, and engagement metrics—to generate a prioritized list of pages for review.

I chose this approach because it bridges specific business decision-making with data analysis and machine learning techniques. Preliminary findings indicate that machine learning models are more effective at ranking pages showing signs of performance decline than simple manual rules; therefore, I believe this approach warrants further exploration.

In [1]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
# My lane: Refresh / Content Opportunity Scoring



## 2. The question: decision, action, cost of a wrong call

### Research question:
Which content pages should be reviewed first for potential refresh based on
observable search, content, and engagement signals?

### Units of content:
One content page

### Desicion:
The decision to be made is which pages the content or SEO team should prioritize
reviewing

### Output:
The output will be a ranked list of pages, with an opportunity score and simple
reason codes explaining why each page was ranked highly

### Action:
A content or SEO reviewer can use the ranked list to decide where to spend review
time first. After reviewing a page, the reviewer may choose to refresh, expand,
protect, monitor, merge, or leave the page unchanged

### Cost of a wrong recommendation:
A false positive could waste reviewer time on a page that did not need attention.
A false negative could cause the team to miss a page with a meaningful decline or
opportunity. Because reviewer time is limited, I care especially about whether the pages near
the top of the ranking are useful review candidates

In [2]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [3]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
import pandas as pd

url = "https://raw.githubusercontent.com/flyrank-bih/flyrank-ml-internship-starter/main/data/raw/content_refresh_anonymized.csv"
df = pd.read_csv(url)

print(df.shape)
df.head()

(30000, 44)


,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7


In [4]:
declining_pages = (df["trend_direction"] == "down").sum()
declining_pct = declining_pages / len(df) * 100

print("Total pages:", len(df))
print("Declining pages:", declining_pages)
print(f"Declining pages (%): {declining_pct:.2f}%")

Total pages: 30000
Declining pages: 16262
Declining pages (%): 54.21%


In [5]:
declining_with_demand = df[(df["trend_direction"] == "down") & (df["impressions_90d"] >= 100)]
print("Declining pages with at least 100 impressions:",len(declining_with_demand))


Declining pages with at least 100 impressions: 13152


In [6]:
low_ctr_candidates = df[(df["impressions_90d"] >= 500) & (df["avg_position"] > 0) & (df["avg_position"] <= 20) & (df["ctr"] < 0.5)]
print("Visible pages with low CTR:",len(low_ctr_candidates))

Visible pages with low CTR: 9759


The starter dataset contains 30,000 content pages.
I found that 16.262 pages (54.%) are currently classified as having a downward trend.
Among them, 13. pages still received at least 100 impressions in the last 90 days.
I also found 9759 pages with at least 500 impressions, an average position between
1 and 20, and CTR below 0.5%.

## 4. Careful words: what I can and can't claim

### What I can
I can describe patterns observed in the available data and build a decision-support
system that ranks pages for human review.
I can compare a simple baseline with a learned method and evaluate whether the
learned method improves the ranking on held-out data.

### What can't
I cannot claim that the model has discovered Google's ranking algorithm.
I also cannot claim that refreshing a highly ranked page will cause its traffic or
ranking to improve. The data is observational, so a high score means that a page may
be worth reviewing, not that a specific action is guaranteed to work.
The current starter decline label is also only a proxy based on the current trend
window. A stronger later version of the project should use a future outcome window
when the warehouse data is available.

In [7]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.